### ISSUE TITLE
Parameter shadowing

### ISSUE THREAD

**Issue by jackmellis** [rollup/rollup#2990]
### Feature Use Case
[example](https://rollupjs.org/repl/?version=1.16.7&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMGZldGNoTWVzc2FnZXMlMjBmcm9tJTIwJ2ZldGNoLW1lc3NhZ2VzLWFwaSclM0IlNUNuaW1wb3J0JTIwY29tcG9zaXRlJTIwZnJvbSUyMCcuJTJGY29tcG9zaXRlJyUzQiU1Q24lNUNuZXhwb3J0JTIwZGVmYXVsdCUyMGNvbXBvc2l0ZShmZXRjaE1lc3NhZ2VzKSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMmZvby5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjBmZXRjaE1lc3NhZ2VzJTIwZnJvbSUyMCdmZXRjaC1tZXNzYWdlcyclM0IlNUNuJTVDbmNvbnN0JTIwZm9vJTIwJTNEJTIwKCklMjAlM0QlM0UlMjBmZXRjaE1lc3NhZ2VzKCklM0IlNUNuJTVDbmV4cG9ydCUyMGRlZmF1bHQlMjBmb28lM0IlMjIlN0QlMkMlN0IlMjJuYW1lJTIyJTNBJTIyYmFoLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmNvbnN0JTIwYmFoJTIwJTNEJTIwKGZldGNoTWVzc2FnZXMpJTIwJTNEJTNFJTIwZmV0Y2hNZXNzYWdlcygpJTNCJTVDbiU1Q25leHBvcnQlMjBkZWZhdWx0JTIwYmFoJTNCJTIyJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMmNvbXBvc2l0ZS5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJpbXBvcnQlMjBmb28lMjBmcm9tJTIwJy4lMkZmb28nJTNCJTVDbmltcG9ydCUyMGJhaCUyMGZyb20lMjAnLiUyRmJhaCclM0IlNUNuJTVDbmNvbnN0JTIwY29tcG9zaXRlJTIwJTNEJTIwKGZldGNoTWVzc2FnZXMpJTIwJTNEJTNFJTIwJTdCJTVDbiUyMCUyMHJldHVybiUyMCU3QiU1Q24lNUN0JTVDdGZvbyUzQSUyMGZvbygpJTJDJTVDbiU1Q3QlNUN0YmFoJTNBJTIwYmFoKGZldGNoTWVzc2FnZXMpJTJDJTVDbiU1Q3QlN0QlM0IlNUNuJTdEJTNCJTVDbiU1Q25leHBvcnQlMjBkZWZhdWx0JTIwY29tcG9zaXRlJTNCJTIyJTdEJTVEJTJDJTIyb3B0aW9ucyUyMiUzQSU3QiUyMmZvcm1hdCUyMiUzQSUyMmVzbSUyMiUyQyUyMm5hbWUlMjIlM0ElMjJteUJ1bmRsZSUyMiUyQyUyMmFtZCUyMiUzQSU3QiUyMmlkJTIyJTNBJTIyJTIyJTdEJTJDJTIyZ2xvYmFscyUyMiUzQSU3QiU3RCU3RCUyQyUyMmV4YW1wbGUlMjIlM0FudWxsJTdE)

### Feature Proposal
basically it would be really useful if rollup avoided variable shadowing as currently this is breaking when I run terser in a large webpack project. I'm pretty sure the problem is deep within terser, but it wouldn't occur at all if rollup could rename conflicting parameters

**Comment by lukastaegert** [comment]
I'm sorry but could you point out where the generated code is wrong? Usually, Rollup successfully deshadows parameters where necessary, e.g. https://rollupjs.org/repl/?version=1.16.7&shareable=JTdCJTIybW9kdWxlcyUyMiUzQSU1QiU3QiUyMm5hbWUlMjIlM0ElMjJtYWluLmpzJTIyJTJDJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMCU3QmZvbyUyMGFzJTIwYmFyJTdEJTIwZnJvbSUyMCcuJTJGZm9vJyU1Q24lNUNuZnVuY3Rpb24lMjBkaXNwbGF5Rm9vKGZvbyklMjAlN0IlNUNuJTVDdGNvbnNvbGUubG9nKGZvbyUyQyUyMGJhciklM0IlNUNuJTdEJTVDbiU1Q25kaXNwbGF5Rm9vKDEpJTNCJTVDbiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTdEJTJDJTdCJTIybmFtZSUyMiUzQSUyMmZvby5qcyUyMiUyQyUyMmNvZGUlMjIlM0ElMjJleHBvcnQlMjBjb25zdCUyMGZvbyUyMCUzRCUyMDElM0IlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXNtJTIyJTJDJTIybmFtZSUyMiUzQSUyMm15QnVuZGxlJTIyJTJDJTIyYW1kJTIyJTNBJTdCJTIyaWQlMjIlM0ElMjIlMjIlN0QlMkMlMjJnbG9iYWxzJTIyJTNBJTdCJTdEJTdEJTJDJTIyZXhhbXBsZSUyMiUzQW51bGwlN0Q=

but I do not see where this is necessary in the example

**Comment by jackmellis** [comment]
In my example (in the ESM view) you can see on line 6 and line 8 we've shadowed the `fetchMessages` variable. This shadowing was done by rollup concatenating the files, not by our own code.

The issue we were having was when Terser was attempting to inline some of our code, it was not accounting for parameter shadowing, and so the entire build was breaking. (Our current solution is to just disable inlining in Terser entirely).

I opened an issue with terser for this but I think if rollup was able to rename all shadowed parameters as well, it'd make this problem a lot easier to resolve. (Hence why I opened a feature not a bug).

Sorry for the delayed reply
